### PR TITLE
optimize copy_resource/5 using dictionaries and make it default resource shape validator

### PR DESCRIPTION
**Pull request by likelion**
Addresses corresponding improvement from #20.

### CODE REVIEW

**Comment by likelion** [comment]
I think this one is complete now. `copy_resource/5` checks the source resource for shape compliance and copies it. The destination resource is not checked as filters may be applied during copying that violate some shape, but this is the desired behaviour.

Current assumption is that none of the shapes describing a resource have overlapping properties, an error is thrown if overlap is detected (see discussion at http://dev.eclipse.org/mhonarc/lists/lyo-dev/msg00883.html).

This implementation should increase performance as it reduces number of triple store accesses by remembering shape substructures in dicts. Further improvement may be to cache dicts but it is not obvious that this will greatly improve the performance.

File: oslc_prolog/applications/oslc_cliopatria_ui.pl [hunk 1]
```diff
@@ -0,0 +1,225 @@
+:- module(oslc_cliopatria_ui, []).
+
+:- http_handler(root(oslc_resource_shapes), oslc_resource_shapes, []).
+:- http_handler(root(oslc_shape_resources), oslc_shape_resources, []).
+:- http_handler(root(oslc_shape), oslc_shape, []).
+
+:- include(bundle(html_page)).
+:- use_module(components(label)).
+
+cliopatria:menu_item(300=places/oslc_resource_shapes, 'Shapes').
```
Comment (berezovskyi) [inline oslc_prolog/applications/oslc_cliopatria_ui.pl:10]: Is 300 an initial dict size?

File: oslc_prolog/applications/oslc_cliopatria_ui.pl [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    td(style('text-align:center'), a(href(ShapeResourcesURI), ResourceCount))
+  ]).
+
+%--------------- RESOURCES DESCRIBED BY SHAPE
+
+oslc_shape_resources(Request) :-
+  http_parameters(Request, [ s(Shape, []) ]),
+  findall(R, rdf(R, oslc:instanceShape, Shape), Resources),
+  sort(Resources, SortedResources),
+  reply_html_page(cliopatria(oslc),
```
Comment (berezovskyi) [inline oslc_prolog/applications/oslc_cliopatria_ui.pl:75]: 😂

**Review by berezovskyi — APPROVED** [review]

**Comment by berezovskyi** [comment]
I was able to run the patch and get to the shape endpoints but unable to do much on them, esp. can't figure out how to reach to the core OSLC services you are providing:

image

File: oslc_prolog/applications/oslc_cliopatria_ui.pl (hunk 1 above)
Comment (likelion) [reply oslc_prolog/applications/oslc_cliopatria_ui.pl:10]: No, 300 is the weight specifying where to place the 'Shapes' menu in the menu bar. :)

File: oslc_prolog/applications/oslc_cliopatria_ui.pl (hunk 2 above)
Comment (likelion) [reply oslc_prolog/applications/oslc_cliopatria_ui.pl:75]: Pretty declarative, isn't it?

**Comment by likelion** [comment]
I'm not providing any service, this UI is to browse resource shapes declared in the triple store ("Places"->"Shapes") and to see what properties shapes describe (click on a number in "Properties" column) and what existing resources are there in the triple store that the shapes describe (click on a number in the "Resources" column).

**Pull request merged**